### ISSUE TITLE
duplicate STATSGO geometry returned by SDA

### ISSUE THREAD

**Issue by dylanbeaudette** [ncss-tech/soilDB#143]
Noticed this problem after adding an SDA-based area calculation. Will have to ask Phil about this.

```r
library(soilDB)
library(sp)
library(digest)
library(raster)
library(rgeos)

b <- c(-119.747629, -119.67935, 36.912019, 36.944987)

# convert vector of corners to polygon for later conversion to WKT by SDA_spatialQuery()
bbox.sp <- as(extent(b), 'SpatialPolygons')
proj4string(bbox.sp) <- '+proj=longlat +datum=WGS84'

# convert to AEA for area calculation
bbox.aea <- spTransform(bbox.sp, CRSobj = CRS('+proj=aea +datum=WGS84 +lat_0=23 +lon_0=-96 +lat_1=29.5 +lat_2=45.5 +x_0=0 +y_0=0  +units=m +no_defs'))

# convert m^2 -> ac.
(bbox.area.ac <- gArea(bbox.aea) * 0.000247105)

# could also do it this way: same result
# raster::area(bbox.sp) * 0.000247105

# return SSURGO polygons, after intersection with provided BBOX
ssurgo.geom <- SDA_spatialQuery(
  bbox.sp, 
  what = 'geom', 
  db = 'SSURGO', 
  geomIntersection = TRUE
)

# return STATSGO polygons, after intersection with provided BBOX
statsgo.geom <- SDA_spatialQuery(
  bbox.sp, 
  what = 'geom', 
  db = 'STATSGO', 
  geomIntersection = TRUE
)

# overlapping vs. intersecting
statsgo.geom.2 <- SDA_spatialQuery(
  bbox.sp, 
  what = 'geom', 
  db = 'STATSGO', 
  geomIntersection = FALSE
)


# this is correct
sum(ssurgo.geom$area_ac)

# this is 2x correct value, why?
sum(statsgo.geom$area_ac)

# investigate duplicate features
featureHash <- function(x) {
  
  # extract polygons
  pp <- lapply(slot(x, 'polygons'), slot, 'Polygons')
  
  # assuming that all polygons slots contain a single polygon, ok
  if(!all(sapply(pp, length) == 1))
    message('multipart polygons, using first polygon')
  
  # compute unique hash for each polygon
  d <- sapply(pp, function(i) {
    # round coordinates (GCS / WGS84) to 4 decimal places
    xy <- round(slot(i[[1]], 'coords'), 4)
    # hash
    digest(xy, algo = 'murmur')
  })
  
  # save hash
  x$hash <- d
  
  # re-order
  x <- x[order(x$hash), ]
  
  return(x)
}


# compute feature hashes
ssurgo.intersection <- featureHash(ssurgo.geom)
statsgo.intersection <- featureHash(statsgo.geom)
statsgo.overlap <- featureHash(statsgo.geom.2)

# eval
# no dupes
any(table(ssurgo.intersection$hash) > 1)
nrow(ssurgo.intersection)

# dupes, 2x each feature
any(table(statsgo.intersection$hash) > 1)
table(statsgo.intersection$hash)
nrow(statsgo.intersection)

# dupes, not by the hash method
# duplicates are very subtly different
any(table(statsgo.overlap$hash) > 1)
table(statsgo.overlap$hash)
nrow(statsgo.overlap)


par(mfrow = c(3,4))
for(i in 1:12) {
  plot(ssurgo.intersection[i, ])
  mtext(ssurgo.intersection$hash[i], side = 1, cex = 0.8, font = 2)
}

par(mfrow = c(3,4))
for(i in 1:nrow(statsgo.intersection)) {
  plot(statsgo.intersection[i, ])
  mtext(statsgo.intersection$hash[i], side = 1, cex = 0.8, font = 2)
  mtext(i, side = 1, line = 2)
}

par(mfrow = c(3,4))
for(i in 1:nrow(statsgo.overlap)) {
  plot(statsgo.overlap[i, ])
  mtext(statsgo.overlap$hash[i], side = 1, cex = 0.8, font = 2)
  mtext(i, side = 1, line = 2)
}
```

![image](https://user-images.githubusercontent.com/624277/93399513-95975e80-f832-11ea-9956-a312114028bb.png)

An example query.
```sql
WITH geom_data (geom, mukey) AS (
  SELECT 
  mupolygongeo.STIntersection( geometry::STGeomFromText('POLYGON ((-119.7476290000000034 36.9120190000000008, -119.7476290000000034 36.9449869999999976, -119.6793499999999995 36.9449869999999976, -119.6793499999999995 36.9120190000000008, -119.7476290000000034 36.9120190000000008))', 4326) ) AS geom, P.mukey
  FROM gsmmupolygon AS P
  WHERE mupolygongeo.STIntersects( geometry::STGeomFromText('POLYGON ((-119.7476290000000034 36.9120190000000008, -119.7476290000000034 36.9449869999999976, -119.6793499999999995 36.9449869999999976, -119.6793499999999995 36.9120190000000008, -119.7476290000000034 36.9120190000000008))', 4326) ) = 1
)
SELECT 
geom.STAsText() AS geom, mukey,
GEOGRAPHY::STGeomFromWKB(
    geom.STUnion(geom.STStartPoint()).STAsBinary(), 4326).STArea() * 0.000247105 AS area_ac
FROM geom_data;
```

**Comment by dylanbeaudette** [comment]
This just in from Phil: there are two versions of STATSGO, must filter using:

```sql
WHERE CLIPAREASYMBOL = 'US'
```

**Comment by dschlaep** [comment]
This seems downstream from the PR https://github.com/ncss-tech/soilDB/issues/143 that I proposed. I noticed the duplication as well and, instead of asking, I eliminate it for the unit test (see the `unique` that I added to the new unit tests in line 123 of https://github.com/ncss-tech/soilDB/pull/142/files#diff-db23cad7472dbf3efe95feaabae48142); I'm sorry.

Once this issue is addressed, then that unit test with db = "STATSGO" could be simplified to how the case db = "SSURGO" is tested.

**Comment by dylanbeaudette** [comment]
Should be fixed with the next commit(7949e7518d27ba55a901e3b8a1cc1c10138e2390) via new internal function `.SDA_geometrySelector`. @dschlaep, I modified the test as you suggested.

Please try to break the "fixed" version of this function!
